src/Workspaces/Core/Portable/Classification/Classifier.cs

```diff
@@ -28,131 +47,29 @@ public static class Classifier
- 
-            return await GetClassifiedSymbolDisplayPartsAsync(
-                semanticModel, textSpan,
-                document.Project.Solution.Workspace,
-                insertSourceTextInGaps,
-                cancellationToken).ConfigureAwait(false);
-        }
- 
-        internal static async Task<List<SymbolDisplayPart>> GetClassifiedSymbolDisplayPartsAsync(
-            SemanticModel semanticModel, TextSpan textSpan, Workspace workspace,
-            bool insertSourceTextInGaps = false,
-            CancellationToken cancellationToken = default(CancellationToken))
-        {
-            var classifiedSpans = GetClassifiedSpans(semanticModel, textSpan, workspace, cancellationToken);
-            var sourceText = await semanticModel.SyntaxTree.GetTextAsync(cancellationToken).ConfigureAwait(false);
-
-            return ConvertClassifications(sourceText, textSpan.Start, classifiedSpans, insertSourceTextInGaps);
-        }
- 
-        internal static List<SymbolDisplayPart> ConvertClassifications(
-            SourceText sourceText, int startPosition, IEnumerable<ClassifiedSpan> classifiedSpans, bool insertSourceTextInGaps = false)
-        {
-            var parts = new List<SymbolDisplayPart>();
- 
-            foreach (var span in classifiedSpans)
+            var service = workspace.Services.GetLanguageServices(semanticModel.Language).GetService<ClassificationService>() as CommonClassificationService;
+            if (service != null)
             {
-                // If there is space between this span and the last one, then add a space.
-                if (startPosition != span.TextSpan.Start)
+                var syntacticClassifications = service.GetSyntacticClassifications(semanticModel.SyntaxTree, textSpan, workspace, cancellationToken);
+                var semanticClassifications = service.GetSemanticClassifications(semanticModel, textSpan, workspace, cancellationToken);
+
+                var list = SharedPools.Default<List<ClassifiedSpan>>().Allocate();
+                try
                 {
-                    if (insertSourceTextInGaps)
-                    {
-                        parts.Add(new SymbolDisplayPart(SymbolDisplayPartKind.Text, null,
-                            sourceText.ToString(TextSpan.FromBounds(
-                                startPosition, span.TextSpan.Start))));
-                    }
-                    else
-                    {
-                        parts.AddRange(Space());
-                    }
+                    list.AddRange(semanticClassifications.Where(cs => cs.TextSpan.OverlapsWith(textSpan)));
+                    var semanticSet = semanticClassifications.Select(cs => cs.TextSpan).ToImmutableHashSet();
+                    list.AddRange(syntacticClassifications.Where(cs => cs.TextSpan.OverlapsWith(textSpan) && !semanticSet.Contains(cs.TextSpan)));
+                    list.Sort((s1, s2) => s1.TextSpan.Start - s2.TextSpan.Start);
```

Comment:
these lines look nearly identical to the function above.  Can we share these somewhere?